src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs

```diff
@@ -90,25 +97,89 @@ public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache? lengthC
-            var clrHandlers = new CompositeMemberHandler<T>[pgFields.Count];
-            var clrHandlerCount = 0;
 
-            var clrHandlerType = IsValueType<T>.Value
+            var clrType = typeof(T);
+            var clrMemberHandlers = new CompositeMemberHandler<T>[pgFields.Count];
+            var clrMemberHandlerCount = 0;
+            var clrMemberHandlerType = IsValueType<T>.Value
                 ? typeof(CompositeStructMemberHandler<,>)
                 : typeof(CompositeClassMemberHandler<,>);
+            var clrConstructorHandlerType = typeof(CompositeParameterHandler<>);
+
+            ConstructorInfo? clrDefaultConstructor = null;
+            CompositeConstructorHandler<T>? clrConstructorHandler = null;
+
+            foreach (var clrConstructor in clrType.GetConstructors())
+            {
+                var clrParameters = clrConstructor.GetParameters();
+                if (clrParameters.Length != pgFields.Count)
+                {
+                    if (clrParameters.Length == 0)
+                        clrDefaultConstructor = clrConstructor;
+
+                    continue;
+                }
+
+                var clrParameterHandlers = new CompositeParameterHandler[pgFields.Count];
+                var clrParameterHandlerCount = 0;
+
+                foreach (var clrParameter in clrParameters)
+                {
+                    var attr = clrParameter.GetCustomAttribute<PgNameAttribute>();
+                    var name = attr?.PgName ?? _nameTranslator.TranslateMemberName(clrParameter.Name!);
+
+                    for (var pgFieldIndex = pgFields.Count - 1; pgFieldIndex >= 0; --pgFieldIndex)
+                    {
+                        var pgField = pgFields[pgFieldIndex];
+                        if (pgField.Name != name)
+                            continue;
+
+                        if (clrParameterHandlers[pgFieldIndex] != null)
+                            throw new AmbiguousMatchException($"Multiple constructor parameters are mapped to the '{pgField.Name}' field.");
+
+                        if (!_typeMapper.TryGetByOID(pgField.Type.OID, out var handler))
+                            throw new NpgsqlException($"PostgreSQL composite type {pgType.DisplayName} has field {pgField.Type.DisplayName} with an unknown type (OID = {pgField.Type.OID}).");
+
+                        clrParameterHandlerCount++;
+                        clrParameterHandlers[pgFieldIndex] = (CompositeParameterHandler)Activator.CreateInstance(
+                            clrConstructorHandlerType.MakeGenericType(clrParameter.ParameterType),
+                            BindingFlags.Instance | BindingFlags.Public,
+                            binder: null,
+                            args: new object[] { handler, clrParameter },
+                            culture: null)!;
+
+                        break;
+                    }
+                }
+
+                if (clrParameterHandlerCount == pgFields.Count)
```

Comment:
Consider using labels and goto on the different loops rather than clrParameterHandlerCount and this check - it may improve readability (not sure).

Reply:
It won't but making a local method could. Will do.